### DISCUSSION TITLE
Using multiple loggers in the config file when training

### DISCUSSION THREAD

**Discussion by farrandi** [explosion/spaCy#9562] [Help: Coding & Implementations] [answered]
I made a custom logger which I added to the config.cfg file:
`@loggers = "mlflow_logger.v1"`

It works and logs the values to mlflow but it does not show the results in the console. I want spacy to both log it to mlflow and print it to the console. **Is there a way I could do this?**

I tried logging 2 loggers by:
`@loggers = "spacy.ConsoleLogger.v1",  "mlflow_logger.v1"`

but it did not work, gave me this error:
`catalogue.RegistryError: [E893] Could not find function '"spacy.ConsoleLogger.v1", "mlflow_logger.v1"' in function registry 'loggers'.`

**Comment by polm** [comment, marked as answer]
I don't believe there's a way to specify multiple loggers at the moment. What you can do is write [a custom logger](https://spacy.io/usage/training#custom-logging) that sends data to both the console and the mlflow logger.

**Reply by farrandi** [reply]
Sorry for the late reply, I tried doing it and it works However, I do have some formatting problems:
```

=========================== Initializing pipeline ===========================
[2021-11-12 12:02:52,862] [INFO] Set up nlp object from config
[2021-11-12 12:02:52,870] [INFO] Pipeline: ['sentencizer', 'tok2vec', 'ner']
[2021-11-12 12:02:53,145] [INFO] Created vocabulary
[2021-11-12 12:02:53,146] [INFO] Finished initializing nlp object
[2021-11-12 12:02:58,148] [INFO] Initialized pipeline components: ['tok2vec', 'ner']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['sentencizer', 'tok2vec', 'ner']
ℹ Frozen components: ['sentencizer']
ℹ Initial learn rate: 0.001
step
score
loss ner


0
0.0
61.69229990243912


200
0.46699015617934536
8290.14936476677
```
It prints something like above in the console, when I want it to be in a table-like format (usual spacy)

This is my current code:
```
@spacy.registry.loggers("mlflow_logger.v1")
def mlflow_logger():
    def setup_logger(
        nlp: Language,
        stdout: IO=sys.stdout,
        stderr: IO=sys.stderr
    )-> Tuple[Callable, Callable]:
        print("step\t")
        print("score\t")
        print("loss ner\t")
        print("\n")
        def log_step(info: Optional[Dict[str, Any]]):
            if info:
                mlflow.log_metric("step", info['step'])
                mlflow.log_metric("score", info['score'])
                mlflow.log_metric(f"loss_ner", info['losses']['ner'])

                print(f"{info['step']}\t")
                print(f"{info['score']}\t")
                print(f"{info['losses']['ner']}\t")
                print("\n")
        def finalize():
            # mlflow.set_terminated(mlflow.active_run().info.run_id)
             print("Model saved in run %s" % mlflow.active_run().info.run_uuid)

        return log_step, finalize

    return setup_logger
```

Is there a way I could see the code in which spacy normally does it?

**Reply by polm** [reply]
Code for the default ConsoleLogger is [here](https://github.com/explosion/spaCy/blob/24cdd4c88e291d01aff9cd92c37c6cdb03479729/spacy/training/loggers.py#L26). 

```
                print(f"{info['step']}\t")
                print(f"{info['score']}\t")
                print(f"{info['losses']['ner']}\t")
                print("\n")
```

This code is your issue - `print` doesn't just write to stdout, it automatically adds a newline at the end of each call. If you combine all your info there into one call to `print` you should get the output you want. You also don't have to add a newline.

**Reply by farrandi** [reply]
I see, thanks!
